docs/proposals/train_api_proposal.md

```diff
@@ -0,0 +1,137 @@
+
+**<h3>New Train API proposal:</h3>**
+
+```python
+From kubeflow.training import trainingClient
+# Arguments related to the model provider including credentials
+model_args = modelProviderClass()
+# Arguments related to the dataset provider including credentials
+dataset_args = datasetProviderClass()
+# Arguments related to the trainer code
+parameters = {key:value pairs}
+trainingClient.train(
+   nodes=1, 
+   nprocs_per_node=4, 
+   model='provider://repo/model_name', 
+   dataset= 'provider://train_dataset_path',
+   eval_dataset = "provider://eval_dataset_path"
+   model_args, 
+   dataset_args, 
+   parameters
+)
+```
+
+Example: 
+
+```python
+@dataclass
+class HuggingFace:
+    access_token = field()
+    peft_config = field()
+    transformerClass = field()
+
+@dataclass
+class S3:
+    access_token = field()
+    region = field()
+ 
+trainingClient.train(
+   nodes=1, 
+   nprocs_per_node=4, 
+   model='hf://openchat/openchat_3.5', 
+   dataset= 's3://doc-example-bucket1/train_dataset',
+   eval_dataset = "s3://doc-example-bucket1/eval_dataset"
+   HuggingFace(access_token = "hf_..." , peft_config = {lora_alpha: 16}, transformerClass="Trainer"), 
+   s3(access_token = "s3 access token", region="us-west-2"), 
+   {learning_rate=0.1}
+)
+```
+
+The new proposed API takes following arguments 
+
+1. System parameters - Number of nodes, number of procs per node(GPUs per node).
+2. Model parameters - Model provider and repository details.
+3. Dataset parameters - Dataset provider and dataset details.
+4. Training parameters - Training specific parameters like learning rate etc.
+
+**<h3>Implementation</h3>**
+
+1. Setup **init** **containers** that download the model and dataset to a PVC. Based on the specified model provider, corresponding training utility functions will be used. Eg: For Huggingface provider, Huggingface trainer can be used. For this **get_pytorchjob_template** function in the sdk needs to be changed to add init containers spec.. Inorder to download models and data sets, we need to support different providers like kaggle, hugging face, s3 or git lfs. The data can be stored in a shared volume between the init container and the main container.
+A new folder containing the code for downloading model and dataset can be added to generate the images for init_containers.
+```
+sdk/python
+        -> kubeflow
+            -> model_init_container_images
+                -> hugging_face.py
+            -> dataset_init_container_images
+                -> s3.py
+```
+
+
+2. Currently, **create_job** api doesn’t support **num_of_nodes** and **gpus_per_node.** We need to add support for that as well, so that the pytorch job with the spec mentioned in[ https://github.com/kubeflow/training-operator/issues/1872#issue comment-1659445716](https://github.com/kubeflow/training-operator/issues/1872#issuecomment-1659445716) can be created.
```

Comment:
```suggestion
2. Currently, **create_job** api doesn’t support **num_of_nodes** and **gpus_per_node.** We need to add support for that as well, so that the pytorch job with the spec mentioned in [https://github.com/kubeflow/training-operator/issues/1872#issue comment-1659445716](https://github.com/kubeflow/training-operator/issues/1872#issuecomment-1659445716) can be created.
```